src/models/gamma_poisson_model.py

```diff
@@ -247,116 +249,317 @@ def _impression_count(self, h):
+            Imax0, alpha0, beta0 = self._fit_histogram_fixed_N(
+                point, N0, one_plus_reach_weight
+            )
+
+            def gamma_obj(params):
+                """Objective function for optimization."""
+                alpha, beta, N = params
+                return self._chi2_distance_with_one_plus_reach(
+                    h, I0, N, alpha, beta, one_plus_reach_weight
+                )
+
+            result = scipy.optimize.minimize(
+                gamma_obj,
+                (alpha0, beta0, N0),
+                method="L-BFGS-B",
+                bounds=((1e-6, 100.0), (1e-6, 100.0), (Nmin, 100.0 * Nmin)),
+            )
+
+            if result.fun < best_score:
+                best_alpha, best_beta, best_N = result.x
+                best_score = result.fun
+
+            if best_score < early_stopping_value:
+                break
+
+        best_Imax = self._expected_impressions(best_N, best_alpha, best_beta)
+        return best_Imax, best_N, best_alpha, best_beta
+
+    def _find_starting_point_for_fixed_N(self, h, I0, N, one_plus_reach_weight):
+        """Searches for a good initial starting point for optimization.
+
+        The standard optimizers unfortunately often fail to find the
+        global minimum if they are started from suboptimal starting points.
+        The surface represented by the objective function has large flat
+        regions, and it often contains a valley with a false minimum that
+        is adjacent to the basin containing the true minimum.  However,
+        it seems to be the case that the basin containing the true minimum
+        is often located relatively close to the origin.  This function performs
+        a simple downhill search for a local optimum starting from a point
+        close to the origin.  The value returned by this function can then
+        be used as the starting point for a more thorough optimization
+        attempt.
+
+        Args:
+          h:  Histogram with kplus reach bucket for the reach point
+              that is to be fit.
+          I0: The number of impressions that were shown to users.
+          N:  The total audience size.
+          one_plus_reach_weight:  Weight to be given to the squared
+            difference in 1+ reach values for the actual and estimated
+            histograms.
+        Returns:
+          alpha, beta: The values of the parameters that gave the best
+          score of the objective function when performing a simple
+          downhill search.
+        """
 
-    def _fit_histogram_fixed_N(self, h, N):
-        """Chi-squared fit to histogram h for fixed N.
+        # Search for local optimum on a grid whose initial spacing is
+        # given by grid_width
+        grid_width = 0.05
+
+        # Try to start from a point that is near the origin but also
+        # in the feasible region
+        alpha = beta = 0.1
+        if I0 > N:
+            alpha = beta = 1.1 * np.sqrt(I0 / N - 1) + grid_width
```

Comment:
The expected total number of impressions for N users is given by the formula:
   E[I] = N * (alpha * beta + 1)

In other words, E[I] / N - 1 = alpha * beta.  

In our case, I0 is the actual number of impressions that were shown to users.  This value has to be less than the total impression inventory, so I0 < E[I].  Thus, we have:

alpha * beta >= I0/N - 1.

Setting alpha = beta, we have alpha^2 >= I0/N - 1, or alpha >= np.sqrt(I0/N - 1).

The purpose of multiplying by 1.1 and adding grid_width was to ensure that the starting point is far enough away from this minimum value to avoid getting NaN values.  Probably we could delete the 1.1 factor, but the algorithm seems to work as it is.

Reply:
This is cool!